tests/python/tenpy/gates/basic_gates_test.py

```diff
@@ -0,0 +1,288 @@
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Tests for the TeNPy basic gates."""
```

Comment:
I see that in the gate tests, you do the following:
1. Apply the gate to a product state as a state vector
2. Apply the gate to a product state as an MPS
3. Check that the expectation value on a molecular Hamiltonian matches in both cases.

Is it possible to retrieve the state vector amplitudes from the MPS? If so, then we don't need to introduce a Hamiltonian. We can just directly check that amplitudes stored by the MPS match the exact ones.

Another question is whether we can apply the evolution not to a product state, but to a random state instead. The basic gates often have trivial action on a product state. For example, applying a Givens rotation between two occupied orbitals doesn't do anything.

Reply:
It is possible to convert an MPS to statevector using the [mps_to_full](https://tenpy.readthedocs.io/en/latest/reference/tenpy.algorithms.exact_diag.ExactDiag.html#tenpy.algorithms.exact_diag.ExactDiag.mps_to_full) function. Yes, I agree that we should remove the MolecularHamiltonian from these gate tests, and keep the tests for the molecular Hamiltonian separate.

Generating a completely random MPS is non-trivial, since a generic random state from Hilbert space will have volume law entanglement, whereas MPS are good at representing states with area law entanglement. I presume that you mean a completely random state, and not just a random product state? One solution is to start with a random product state and then perform a TEBD [RandomUnitaryEvolution](https://tenpy.readthedocs.io/en/latest/reference/tenpy.algorithms.tebd.RandomUnitaryEvolution.html#tenpy.algorithms.tebd.RandomUnitaryEvolution) to introduce some short-range entanglement. This will be not result in a completely random state but it will be entangled to as much as the computational cost / bond dimension will allow.

Reply:
A uniformly random state would be ideal, but perhaps unnecessary. We should be able to represent such a state for small test sizes though, right? You can generate a uniformly random state vector using `ffsim.random.random_state_vector`. Is it possible to convert such a state vector into an MPS using Tenpy? Otherwise, I think your idea of generating a random evolution should be fine.

Reply:
Yes, for small system sizes you may be able to convert a completely random statevector into an MPS using the [full_to_mps](https://tenpy.readthedocs.io/en/latest/reference/tenpy.algorithms.exact_diag.ExactDiag.html#tenpy.algorithms.exact_diag.ExactDiag.full_to_mps) function. However, the bond dimension will blow up quite quickly. I will try in the first instance and if it's not tractable, I will try the random unitary evolution approach.